s3datarepository/s3datarepository_test.go

```diff
@@ -0,0 +1,429 @@
+package s3datarepository
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/YaleSpinup/ds-api/apierror"
+	"github.com/YaleSpinup/ds-api/dataset"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+)
+
+// mockS3Client is a fake S3 client
+type mockS3Client struct {
+	s3iface.S3API
+	t         *testing.T
+	err       map[string]error
+	headCount uint
+}
+
+func newMockS3Client(t *testing.T) s3iface.S3API {
+	return &mockS3Client{
+		t:         t,
+		err:       make(map[string]error),
+		headCount: 0,
+	}
+}
+
+func (m *mockS3Client) HeadBucketWithContext(ctx context.Context, input *s3.HeadBucketInput, opts ...request.Option) (*s3.HeadBucketOutput, error) {
+	if m.err["HeadBucketWithContext"] != nil {
+		if m.headCount == 0 {
+			m.headCount++
+			return nil, m.err["HeadBucketWithContext"]
+		}
+	}
```

Comment:
this isn't wrong, but these should probably use the form:

```golang
if err, ok := m.err["HeadBucketWithContext"]; ok {
....
}
```

when getting the value of a nonexistent key in a map, you always get the zero value of the type... here the zero value of error is `nil`, but it's more confusing IMO than using the “comma ok” idiom.